File: lua/p2bot.lua
require("socket.http")
require("ltn12")
require("io")

function debug_table(table)
    for k, v in pairs(table) do
        print(k .. ':' .. v)
    end
end

function p2db_get(id)
    local data = {}
    local f = io.open("db/" .. id .. ".dat", "r")
    if (not f) then
        return nil
    end

    local s = f:read()
    local id, name, num, modified, length = string.match(s, "(.+)<>(.+)<>(.+)<>(.+)<>(.+)")
    data['id'] = id
    data['name'] = name
    data['num'] = num
    data['modified'] = modified
    data['length'] = length
    f:close()
    return data
end

function p2db_set(id, name, num, modified, length)
    local f = io.open("db/" .. id .. ".dat", "w")
    f:write(string.format("%s<>%s<>%s<>%s<>%s", id, name, num, modified, length))
    f:close()
end

p2db = {}
function p2db.new()
    return {
        get = p2db_get,
        set = p2db_set
    }
end

function http_get(url, headers)
    local body = {}
    local r, s, h = socket.http.request {
        method = "GET",
        url = url,
        sink = ltn12.sink.table(body),
        headers = headers,
        redirect = true
    }
    return table.concat(body), s, h
end

function get_dats()
    local subject = http_get("http://" .. server .. "/subject.txt", { referer = "http://www.google.co.jp/" }) 
    local dats = {}
    for k, v in string.gmatch(subject, "(.-).dat<>(.-)\n") do
        table.insert(dats, { id = k, url = "http://" .. server .. "/dat/" .. k .. ".dat", name = v });
    end
    return dats;
end

-- pl-loader
function download_image(url, outfile)
    print(url)
    --local data, status, header = http_get("http://" .. url)
    --if (status == 200 and header['content-type'] == 'image/jpeg') then
    --    local f = io.open(outfile, "w")
    --    f:write(data)
    --    f:close()
    --end
end

-- If-Modified-Since: Fri, 23 Oct 2009 09:37:06 GMT
-- Range: bytes= 86641-
-- 206 Partial Content
-- 304 Not Modified
-- 203 dat oti
function get_thread(dat)
    local data = db.get(dat['id'])
    local header = {}
    local num = 0

    if data then
        num = data['num']
        header['If-Modified-Since'] = data['modified']
        header['Range'] = "bytes= " .. data['length'] .. "-"
    end
    local html, status, header = http_get(dat['url'], header)
    if (status == 200 or status == 206) then
        for image in string.gmatch(html, "//([%w%d%%%.-^~=?/_:]+%.jpg)") do
            download_image(image, 'images/' .. dat['id'] .. '-' .. num ..'.jpg')
            num = num + 1
            usleep(500000)
        end
        local length = header['content-length']
        if data then
            length = length + data['length']
        end
        db.set(dat['id'], dat['name'], num, header['last-modified'], length)
    end
end

server = "yutori7.2ch.net/news4vip"
socket.http.USERAGENT = "Mozilla/5.0 (compatible; p2bot/0.1; +http://p2m.giox.org/)"
-- http.USERAGENT = "Mozilla/5.0"
db = p2db.new()
dats = get_dats()
for k, v in pairs(dats) do
    get_thread(v)
    collectgarbage("collect")
    sleep(1)
end

File: ruby/p2bot.rb
require 'uri'
require 'socket'

class HTTP
  def get(url, header)
    uri = URI.parse(url)
    s = TCPSocket.new(uri.host, uri.port)
    s.write("GET #{uri.path} HTTP/1.1\r\n")
    s.write("Host: #{uri.host}\r\n")
    s.write("Useer-Agent: Mozilla/5.0\r\n")
    s.write("Connection: close\r\n")
    # "Mozilla/5.0 (compatible; p2bot/0.1; +http://p2m.giox.org/)"
    # Refere
    s.write("\r\n")

    return nil unless s.gets() =~ /(\d{3})/
    status = $1
    header = read_header(s)
    yield(status.to_i, header, s)
    s.close
  end

  def download(url, header, outfile)
    uri = URI.parse(url)
    s = TCPSocket.new(uri.host, uri.port)
    s.write("GET #{uri.path} HTTP/1.1\r\n")
    s.write("Host: #{uri.host}\r\n")
    s.write("Useer-Agent: Mozilla/5.0\r\n")
    s.write("Connection: close\r\n")
    # "Mozilla/5.0 (compatible; p2bot/0.1; +http://p2m.giox.org/)"
    # Refere
    s.write("\r\n")

    return nil unless s.gets() =~ /(\d{3})/
    return nil unless $1 == "200"
    header = read_header(s)
    open(outfile, "w") do |f|
      while (buf = s.read(1024))
        f.write(buf)  
      end
    end
    s.close
  rescue
    p url
  end

  def read_header(socket)
    header = {}
    while true
      line = socket.gets()
      break if line == "\r\n"
      h = line.split(':')
      header[h[0].downcase!] = h[1].strip!
    end
    return header
  end
end

subject = "http://yutori7.2ch.net/news4vip/subject.txt"

dat_list = []
HTTP.new.get(subject, nil) do |status, header, io|
  while (line = io.gets)
    dat = line.split('<>')
    dat_list.push(dat)
  end
end

dat_list.each do |e|
  HTTP.new.get("http://yutori7.2ch.net/news4vip/dat/#{e[0]}", nil) do |status, header, io|
    id = e[0].split('.')[0]
    puts id
    n = 0
    while (line = io.gets)
      line.scan(%r{//\S+?\.jpg}) do |url|
        HTTP.new.download("http:#{url}", nil, "images/#{id}_#{n}.jpg")
        n += 1
        sleep(0.5)
      end
    end
  end
  GC.start
  sleep(1)
end
